File: loong's code/leetcode/editor/cn/1223. 掷骰子模拟.py
from functools import cache
from typing import List


class Solution:
    def dieSimulator(self, n: int, rollMax: List[int]) -> int:
        MOD = 10 ** 9 + 7

        @cache
        def dfs(i: int, last: int, left: int) -> int:
            if i == 0:
                return 1
            res = 0
            for j, mx in enumerate(rollMax):
                if j != last:
                    res += dfs(i - 1, j, mx - 1)
                elif left:
                    res += dfs(i - 1, j, left - 1)
            return res % MOD

        return sum(dfs(n - 1, j, mx - 1) for j, mx in enumerate(rollMax)) % MOD


print(Solution.dieSimulator(Solution(), n=2, rollMax=[1, 1, 2, 2, 2, 3]))

File: loong's code/leetcode/editor/cn/1223. 掷骰子模拟.go
package main

// 会超时的回溯写法
func dieSimulator(n int, rollMax []int) (ans int) {
	const mod int = 1e9 + 7

	const m = 6
	cache := make([][m][]int, n)
	for i := range cache {
		for j := range cache[i] {
			cache[i][j] = make([]int, rollMax[j])
			for k := range cache[i][j] {
				cache[i][j][k] = -1
			}
		}
	}

	var dfs func(int, int, int) int
	dfs = func(i, last, left int) (res int) {
		if i == 0 {
			return 1
		}
		for j, mx := range rollMax {
			if j != last {
				res += dfs(i-1, j, mx-1)
			} else if left > 0 {
				res += dfs(i-1, j, left-1)
			}
		}
		return res % mod
	}
	for j, mx := range rollMax {
		ans += dfs(n-1, j, mx-1)
	}
	return ans % mod
}
